### PR TITLE
Document OpenColorIO Development Process

### LINKED ISSUE #1124: On Development Process Transparency & Roadmap
Hello,

I'm starting an issue to centralise the discussions in relation to my two comments [here](https://github.com/AcademySoftwareFoundation/OpenColorIO/issues/1100#issuecomment-674340278) and [there](https://github.com/AcademySoftwareFoundation/OpenColorIO/pull/1102#issuecomment-682302942).

# Introduction

With PR #1102, OpenColorIO is about to receive three new grading transforms. I want to clarify right from the beginning, that the discussion in this issue is not about the transforms themselves but how we got there and how we can improve the process to get there because it raised some valid concerns in the industry.

The gist of the concerns being: How can 22K lines of code for a significant feature land in the repo without an issue or a public thread being created first to discuss the feature. Put another way: How it is possible that the community is not aware that contributors to the project are working on such large features before they land in the repo?

The problem is compounded by the fact that the two-weeks rule makes it possible to merge such a large feature somehow easily.

#  Development Process Transparency

Given the context, I'm advocating for more transparency. Fortunately, I'm not alone thinking that what happened is abnormal. I think that significant features should be introduced and discussed openly before the work happens.

OCIO is used virtually by all the VFX industry DCCs, commercial or open-source, and as such is a critical piece of software for the industry and its community. It, thus, deserves uttermost clarity in its development process. I certainly feel abnormal that some of the key contributors have been developing a significant feature in the shadows, pushed it to the light and can have it merged within two weeks. I searched the mailing list, the issues, the TSC notes, the Working Group notes and I could not see anything pertaining to that and even if I failed to find it, it should have been easy in the first place.

Don't get me wrong here, like everybody else, I sincerely appreciate the work Doug, Patrick and Bernard have been doing, OCIO resurrected from the ashes thanks to their efforts. 🎊 

Back to the issue at hand: What can we do to avoid such a situation again?

I'm hoping that the process we define here can be applied to other ASWF projects because OCIO is not the only one affected by the sound of it.

# Roadmap

The immediate answer to the problem is that any significant feature on a critical and stable project like OCIO should be discussed beforehand in an issue or, at the very least, during the early stage of the feature development. Sometimes, a small toy project, a test prototype can be turned into a feature and the idea here is not to prevent such initiatives, quite the opposite.

Creating an issue for the feature gives time to the community to discuss it and decide whether it is appropriate for the project. It is also important because it avoids potentially wasted time and energy: Imagine a contributor spending 2 or 3 weeks producing 20K lines of fine code just to get his work rejected because it is deemed out-of-scope.

The other part of the answer is that anything significant on the project should be tied to a roadmap, I know that roadmaps are two-edged swords, especially when stakeholders have expectations about things being delivered on time. I'm not looking for that but more for a view that gives a high-level summary of what is being worked on the project, akin to an Agile view where one can only see the Epics for the project.

An example I gave is Epic Games' Trello Board: https://trello.com/b/TTAVI7Ny/ue4-roadmap

image

I will shut-up and let people chime in!

Bests,

Thomas

### PR DESCRIPTION
**Pull request by michdolan**
Following discussion on transparency and openness of new feature development, and related TSC discussion, I've added a PROCESS.md document to outline a feature proposal process. The process is based on current practices adhered to by the project with two changes:

1. Push up proposal discussion and issue creation in the timeline where possible. In short, it's up to the contributor when a design is proposed, but accepting the risk of potential redesign requested by the community.
2. Adding accepted proposals to a new Roadmap GH Project for progress tracking and transparency.

All of this is up for discussion of course. All comments/suggestions are welcome and appreciated.

Resolves #1124 

Signed-off-by: Michael Dolan <michdolan@gmail.com>

### CODE REVIEW

**Comment by KelSolaar** [comment]
Hi @michdolan,

Looks great at first glance! It covers pretty much what I was hoping to see. I haven’t checked the roadmap project yet though.

Cheers,

Thomas

File: PROCESS.md
Comment (doug-walker) [inline PROCESS.md:69]: This whole section is out of date and needs to be redone.  I realize it is a copy/paste from the previous document, but I don't think we should merge it without updating it to be current.

File: PROCESS.md
Comment (doug-walker) [inline PROCESS.md:87]: The current GitHub protection settings require two committer approvals before anything may be merged.

File: PROCESS.md
Comment (doug-walker) [inline PROCESS.md:81]: This part should be edited or removed as it overlaps with the previous section.  Let's keep this section focused on the Required Approvals and the previous one on the feature development process.

File: PROCESS.md
Comment (doug-walker) [inline PROCESS.md:99]: This seems to restate what is in point 6 in the previous section, although somewhat differently.  It would be better to only describe the rule in one place rather than two.

File: PROCESS.md
Comment (doug-walker) [inline PROCESS.md:38]: Not totally opposed, but I find the "Roadmap" name potentially misleading for people.  A "roadmap" is generally used to define a release or a set of releases.  The GitHub Project as described here is only going to be populated as new features are proposed and accepted.  There is no clear mapping of these to a given release.  It may also give the impression that the roadmap is more empty or closer to completion than is actually the case.  Perhaps a name such as "Feature development" or something like that would be more clear?

**Review by doug-walker — CHANGES_REQUESTED** [review]

**Comment by michdolan** [comment]
Thanks for the feedback! @doug-walker I will take a pass at revising that section based on your feedback. Great points.

File: PROCESS.md
Comment (michdolan) [reply PROCESS.md:38]: Good point. Renamed to `Feature Development`: https://github.com/AcademySoftwareFoundation/OpenColorIO/projects/3

File: PROCESS.md
Comment (michdolan) [reply PROCESS.md]: Fixed.

File: PROCESS.md
Comment (michdolan) [reply PROCESS.md:64]: I revised the `Required Approval's` section and reduced duplication between it and the `Core Library Changes` section.

File: PROCESS.md
Comment (michdolan) [reply PROCESS.md]: Updated to remove clauses requiring less than two approvers.

File: PROCESS.md
Comment (michdolan) [reply PROCESS.md]: Revised.

File: PROCESS.md
Comment (michdolan) [reply PROCESS.md]: Good point. Revised.

**Review by doug-walker — CHANGES_REQUESTED** [review]

File: PROCESS.md
Comment (michdolan) [reply PROCESS.md:72]: I interpreted your previous third comment as suggesting this should be removed:

> The current GitHub protection settings require two committer approvals before anything may be merged.

Comment (michdolan) [reply PROCESS.md:72]: Nevermind, I misunderstood this comment. But the section you are suggesting adding "Large or significant changes" to doesn't carry any condition other than having two approvers, so what would be the distinction between that and the proposed small change rule?

Comment (doug-walker) [reply PROCESS.md:72]: The distinction is that approvers for large changes cannot have the same affiliation, whereas they could for small changes.  For example, if it's a bug fix authored by Patrick, I should be allowed to be an approver even before the two weeks has elapsed.

Comment (hodoulp) [reply PROCESS.md:72]: Thanks to clearly highlight the critical distinction between small pull requests (i.e. bug fix, etc.) and larges ones (i.e. new feature, etc.). The approval process must authorize a fast-track approach for small pull requests to avoid blocking contributors (even if creator and approver have the same affiliation as described by Doug).  OCIOv2 starts its stabilization phase where bug fixes could pop at anytime needing a fast-track merge (i.e. within few days).

File: PROCESS.md
Comment (hodoulp) [reply PROCESS.md:85]: The owner concept is definitively not clear. If needed, it must be based on active contributions (through pull request, etc.) and / or active involvement (through code review, issue review, etc.) in the designated code base.

File: PROCESS.md
Comment (michdolan) [reply PROCESS.md]: Fixed.

File: PROCESS.md
Comment (michdolan) [reply PROCESS.md]: Good point. If we ever wanted to do this it can be achieved with a `CODEOWNERS.md` file which I believe GH supports for directing approvals, but agree that we're not doing this now so it should be removed.

File: PROCESS.md
Comment (michdolan) [reply PROCESS.md:106]: Reverted this section to previous state

File: PROCESS.md
Comment (michdolan) [reply PROCESS.md:72]: I moved the general two approval description to its own bullet point, followed by two others with the rules based on change scope. I also added additional notes around minimum PR lifetime and giving notice before 2 week merges.

File: PROCESS.md
Comment (doug-walker) [inline PROCESS.md:100]: Minor, but we sometimes make this notice more than 48 hours before the 14 days has elapsed.  To me, the longer notice seems like a good thing but the new rule seems to forbid it (i.e. seems to say it cannot happen before day 12).  Propose replacing with the following sentence:

Notice must be given in the pull request between 2 and 4 days in advance of the intended merge date.

Comment (michdolan) [reply PROCESS.md]: Changed

**Review by doug-walker — APPROVED** [review]

**Review by hodoulp — APPROVED** [review]

**Comment by KelSolaar** [comment]
Was that discussed yesterday in the TSC meeting? I thought it would have made sense to publicize/advertize the document on Slack and/or the mailing list to see if the community was fine!

It is a bit of a chicken and egg situation but I will quote the new document itself:

> All core changes begin with community discussion, which can involve informal conversation in Slack or on ocio-dev, but are ultimately brought to a TSC meeting where consensus can be reached around the proposal's relevance and approach. TSC meetings are public and meeting notes are posted to the OCIO repo within days of a meeting for a permanent record of decisions.

So to go back to some of the last bullet points:

> Large or significant changes must receive both approvals from Committers that are not affiliated with the author or their company.

> If approval does not happen within two weeks of opening a pull request and there are no raised objections or unresolved discussions happening about the feature, a Committer that is affiliated with the author may approve it. At some point, we have to assume that the people who know and care are monitoring the PRs and that an extended period without objections is really assent. Notice must be given in the pull request between 2 and 4 days in advance of the intended merge date. This gives a windows for reviewers to comment or request additional time for review.

Isn't this still empowering a group of affiliated people to steer the project where they want? It is Christmas break, Michael is taking some holidays, and we get a raytracer in OCIO at New Year's Eve, because, well nobody really looked at the project... :) I think that Core changes should not have an escape door.

Besides this point, it looks fine!

**Comment by KelSolaar** [comment]
Damn, sorry to have jumped on that too late! Been hell of a (many) week(s) lately.

**Pull request merged**
